slsa-e2e-rfp/commands/verify-vsa/main.go

```diff
@@ -0,0 +1,153 @@
+	vsaPredicateType = "https://slsa.dev/verification_summary/v1"
+)
+
+// FIMXE(trishankkarthik): Read the public key from disk instead of hardcoding it.
+var (
+	policyVSAPubKey = &signerverifier.SSLibKey{
+		KeyIDHashAlgorithms: []string{"sha256", "sha512"},
+		KeyType:             "ed25519",
+		KeyVal: signerverifier.KeyVal{
+			Public: "1f81c9eedcb9325243706a10cd24cf672e5b6f31cd419a4976ea7b6c166181f9",
+		},
+		Scheme: "ed25519",
+		KeyID:  "64b5e39bbf527b383a758f40501005c470df4d5463c3183486da1a04a0cce755",
+	}
+)
+
+// TODO:
+// [x] Verify the Policy VSA using the public key
+// [x] Check that the artifact name and digest in the Policy VSA matches the artifact on disk
+// [x] Check that the verification passed
+// [x] Check that the verification time is not too long ago in the past
+func main() {
+	verifier, err := signerverifier.NewED25519SignerVerifierFromSSLibKey(policyVSAPubKey)
+	if err != nil {
+		panic(err)
+	}
+
+	envelopeVerifier, err := dsse.NewEnvelopeVerifier(verifier)
+	if err != nil {
+		panic(err)
+	}
+
+	envelopeBytes, err := os.ReadFile(policyVSA)
+	if err != nil {
+		panic(err)
+	}
+
+	envelope := &dsse.Envelope{}
+	if err := json.Unmarshal(envelopeBytes, envelope); err != nil {
+		panic(err)
+	}
+
+	acceptedKeys, err := envelopeVerifier.Verify(context.Background(), envelope)
+	if err != nil {
+		panic(err)
+	}
+	if len(acceptedKeys) != 1 {
+		panic("expected exactly one accepted key")
+	}
+	fmt.Printf("Policy VSA verified by public key with ID %s\n", acceptedKeys[0].KeyID)
+
+	payloadBytes, err := envelope.DecodeB64Payload()
+	if err != nil {
+		panic(err)
+	}
+
+	statement := &v1.Statement{}
+	if err := protojson.Unmarshal(payloadBytes, statement); err != nil {
+		panic(err)
+	}
+	if statement.PredicateType != vsaPredicateType {
+		panic(fmt.Sprintf("expected predicate type %s, got %s", vsaPredicateType, statement.PredicateType))
+	}
+
+	artifactFile, err := os.Open(artifactPath)
+	if err != nil {
+		panic(err)
+	}
+	defer artifactFile.Close()
+
+	hasher := sha512.New()
+	if _, err := io.Copy(hasher, artifactFile); err != nil {
+		panic(err)
+	}
+	artifactHash := hex.EncodeToString(hasher.Sum(nil))
+
+	artifactBase := filepath.Base(artifactPath)
+	found := false
+	for _, subject := range statement.Subject {
+		// If we have an artifact with the same name, we want the same digest.
+		subjectHash := subject.Digest["sha2-512"]
+		if subject.Name == artifactBase {
+			if subjectHash != artifactHash {
+				panic(fmt.Sprintf("artifact hash mismatch: expected %s, got %s", artifactHash, subject.Digest["sha2-512"]))
+			}
+			found = true
+		}
+		// And if we have an artifact with the same hash, we want the same name.
+		if subjectHash == artifactHash {
+			if subject.Name != artifactBase {
+				panic(fmt.Sprintf("artifact name mismatch: expected %s, got %s", artifactBase, subject.Name))
+			}
+			found = true
+		}
+	}
+	// So basically, we guarantee there is only one subject with the same name and digest.
```

Comment:
Well, not one subject, but that all subjects with the same name have the same digest, and those with same digest have the same name.